### ISSUE TITLE
allow multiple PISM instances running simultaneously

### ISSUE THREAD

**Issue by bueler** [pism/pism#176]
I did not want this _suggestion_ to get lost.

This is a hoped-for capability from Bob Fischer.  Presumably other PISM users would also like to have both Greenland and Antarctica simulations running and talking to the same running GCM.  At least potentially.

The technical barrier seems primarily to be issue #175, that is, because of the "globalness" of the PETSc options database.

A regression test of this capability would be appropriate.

**Comment by bueler** [comment]
Ancient.  Any reason not to close?

**Comment by ckhroulev** [comment]
@bueler I still want to make this possible, so I'd keep it open.

**Comment by talbrecht** [comment]
Could you explain a little bit more, what you have in mind, @bueler? I know that some users within PalMod context use heterogeneous slurm jobs to run several PISM instances simultaneously. Maybe we can exchange some experience on this topic?!

**Comment by bueler** [comment]
@talbrecht My mind is empty, sorry.  I don't remember the 2013 conversation about it, but I was just wondering if anyone cared, and if I should take responsibility for an old issue and close it.

**Comment by ckhroulev** [comment]
@talbrecht This issue refers to the desired ability to have *one* program using 2 instances of PISM code (`IceModel`, etc), using different MPI communicators, different tuning parameters and so on. This would be needed to *tightly* (in the software sense) couple PISM to a GCM modeling a planet with more than one glaciated area.

This is not related to SLURM. As far as I know we don't have any experience using heterogeneous SLURM jobs here at UAF.

**Comment by mankoff** [comment]
We are hoping to run multiple PISM simultaneously in the near future. Is this still not possible?

It seems an easy workaround would be to put each PISM in its own Docker (or Singularity or similar on the HPC system) container.

**Comment by talbrecht** [comment]
As far as I know, the MPI-ESM in Hamburg is using two PISM instances simultaneously, one for the northern Hemisphere and another for Antarctica.

**Comment by ckhroulev** [comment]
> We are hoping to run multiple PISM simultaneously in the near future. Is this still not possible?

As I mentioned in a comment above, this is about running two or more instances of PISM simultaneously _in one program_, i.e. when using PISM as a software library. This is possible, but I cannot (yet) guarantee that this always works. I need to ensure that the code processing command-line options always respects option prefixes... and then I will be able to close this issue.

Note, though, that almost all of PISM's settings can be controlled using configuration parameters, so we may be able to run two instances of PISM already.

Do you have a setup you have in mind? I should be able to check if it would work *now* and what needs to be done if it does not.

> It seems an easy workaround would be to put each PISM in its own Docker (or Singularity or similar on the HPC system) container.

That is not necessary: you can build PISM once and run as many copies of it as you need: all parameter choices are set at run time (not compilation time).

**Comment by ckhroulev** [comment]
> It sounds like this won't be an issue in this scenario based on your comment above.

This is correct: no problem with running multiple copies of PISM this way.

**Comment by ClemensSchannwell** [comment]
I will add my two cents here. So at the MPI, we have experimented with running two instances of PISM for the northern and southern hemisphere in a single slurm job at the same time. While this did work as expected it was quite fiddly to determine the best distributions of cpus between the two PISM instances beforehand. This in turn often led to a suboptimal run time and a waste of compute resources because one of the PISM instances was idle for a certain amount of time. Since PISM is never the bottleneck in our coupled climate-ice sheet simulations, we have kind of abandoned this exeperiment at the moment. 
What kind of benefits would the library option bring in comparison to running two instances? I guess it could be called from within another model e..g a climate model and would allow doing the coupling online instead of restart files?

**Comment by ckhroulev** [comment]
> So at the MPI, we have experimented with running two instances of PISM for the northern and southern hemisphere in a single slurm job at the same time.

I'm sure you considered it already, but still, here's something that may be worth thinking about. We know that PISM's scaling with the number of MPI jobs is not perfect, so if a simulation using `N` CPU cores takes `T` seconds, a simulation on `2N` cores will take `T/2 + overhead` seconds. If the overhead is small, then running on `2N` cores may be worthwhile.

In the context of running 2 PISM instances (north and south) on `N` cores each, if these scale up to `2N` cores well running them in sequence may be a reasonable idea _as long as_ the associated overhead is no worse than the CPU time wasted when one of the instances is idle waiting for the other one.

Of course this depends on `N` and grid sizes used by these PISM instances.

> While this did work as expected it was quite fiddly to determine the best distributions of cpus between the two PISM instances beforehand.

This is not surprising. Unfortunately load balancing between two PISM instances is hard no matter how you run PISM: the computational cost of advancing the model from `t` to `t + dt` (here `dt` is the coupling time step consisting of multiple PISM time steps) depends on the state of the model, which makes it hard to predict. The best I can come up with is tuning a setup using a "representative" simulation and hoping that further simulations are similar enough to benefit from this tuning.

> What kind of benefits would the library option bring in comparison to running two instances? I guess it could be called from within another model e..g a climate model and would allow doing the coupling online instead of restart files?

Yes, the benefit would be the ability to control PISM "online" from within another model, mainly to avoid the costs associated with
- reading and writing data passed between models
- model initialization (this way it happens *once* and not every coupling time step).
- managing coupling and model state files

All these costs increase with grid resolution and coupling frequency. As you pointed out, in some cases these inefficiencies may be too small (relative to other costs, anyway) to be worth the effort needed to address them.

**Comment by ClemensSchannwell** [comment]
>This is not surprising. Unfortunately load balancing between two PISM instances is hard no matter how you run PISM: the computational cost of advancing the model from t to t + dt (here dt is the coupling time step consisting of multiple PISM time steps) depends on the state of the model, which makes it hard to predict. The best I can come up with is tuning a setup using a "representative" simulation and hoping that further simulations are similar enough to benefit from this tuning.

Yes, this is what we tried as well. But as you pointed out, it really only works if your model state does not change too much (whatever that means), which is a reasonable assumption for centennial-scale simulations. For our glacial-cycle timescales, we have periodic ice-sheet instabilities and during these periods this approach does not work anymore. 

> Yes, the benefit would be the ability to control PISM "online" from within another model, mainly to avoid the costs associated with
> 
> - reading and writing data passed between models
> - model initialization (this way it happens once and not every coupling time step).
> - managing coupling and model state files
> 
> All these costs increase with grid resolution and coupling frequency. As you pointed out, in some cases these inefficiencies may be too small (relative to other costs, anyway) to be worth the effort needed to address them.
> 

From my point of view this is exactly why it would still be nice to have this feature. Because it would allow you to do "online" coupling via a coupler from the climate model to PISM and vice versa which I believe is the way to go as ice sheets get more and more incorporated into Earth System Model frameworks...

**Comment by ckhroulev** [comment]
> From my point of view this is exactly why it would still be nice to have this feature. Because it would allow you to do "online" coupling via a coupler from the climate model to PISM and vice versa which I believe is the way to go as ice sheets get more and more incorporated into Earth System Model frameworks...

Agreed.
